source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,667 @@
+  } else {
+    // Vector operation
+    const analysis::Type* element_type =
+        result_type->AsVector()->element_type();
+    auto new_vec_const =
+        std::unique_ptr<Const>(new VectorConst(result_type, element_type));
+    uint32_t num_dims = result_type->AsVector()->element_count();
+    std::vector<uint32_t> result_vec =
+        OperateVectors(spec_opcode, num_dims, operands);
+    for (uint32_t r : result_vec) {
+      auto rc = CreateConst(element_type, {r});
+      new_vec_const->AsVectorConst()->AppendComponent(rc.get());
+      BuildInstructionAndAddToModule(std::move(rc), pos);
+    }
+    return BuildInstructionAndAddToModule(std::move(new_vec_const), pos);
+  }
+}
+
+ir::Instruction*
+FoldSpecConstantOpAndCompositePass::BuildInstructionAndAddToModule(
+    std::unique_ptr<Const> c, ir::Module::inst_iterator* pos) {
+  Const* new_const = c.get();
+  uint32_t new_id = ++max_id_;
+  const_val_to_id_[c.get()] = new_id;
+  id_to_const_val_[new_id] = std::move(c);
+  auto new_inst = CreateInstruction(new_id, new_const);
+  if (!new_inst) return nullptr;
+  auto* new_inst_ptr = new_inst.get();
+  *pos = pos->InsertBefore(std::move(new_inst));
+  (*pos)++;
+  def_use_mgr_->AnalyzeInstDefUse(new_inst_ptr);
+  return new_inst_ptr;
+}
+
+std::unique_ptr<Const> FoldSpecConstantOpAndCompositePass::CreateConstFromInst(
+    ir::Instruction* inst) {
+  std::vector<uint32_t> literal_words_or_ids;
+  std::unique_ptr<Const> new_const;
+  for (uint32_t i = 0; i < inst->NumInOperands(); i++) {
+    literal_words_or_ids.insert(literal_words_or_ids.end(),
+                                inst->GetInOperand(i).words.begin(),
+                                inst->GetInOperand(i).words.end());
+  }
+  switch (inst->opcode()) {
+    case SpvOp::SpvOpConstantTrue:
+      literal_words_or_ids.push_back(true);
+      break;
+    case SpvOp::SpvOpConstantFalse:
+      literal_words_or_ids.push_back(false);
+      break;
+    case SpvOp::SpvOpConstantNull:
+    case SpvOp::SpvOpConstant:
+    case SpvOp::SpvOpConstantComposite:
+    case SpvOp::SpvOpSpecConstantComposite:
+      break;
+    default:
+      return nullptr;
+  }
+  return CreateConst(GetType(inst), literal_words_or_ids);
+}
+
+Const* FoldSpecConstantOpAndCompositePass::FindRecordedConst(uint32_t id) {
+  auto iter = id_to_const_val_.find(id);
+  if (iter == id_to_const_val_.end())
+    return nullptr;
+  else
+    return iter->second.get();
+}
+
+uint32_t FoldSpecConstantOpAndCompositePass::FindRecordedConst(const Const* c) {
+  auto iter = const_val_to_id_.find(c);
+  if (iter == const_val_to_id_.end())
+    return 0;
+  else
+    return iter->second;
+}
+
+bool FoldSpecConstantOpAndCompositePass::AddComponentsTo(
+    const std::vector<uint32_t>& ids, CompositeConst* composite_const) {
+  for (uint32_t id : ids) {
+    if (Const* c = FindRecordedConst(id)) {
+      composite_const->AppendComponent(c);
+    } else {
+      return false;
```

Comment:
This will leave the `composite_const` half baked. Is that OK? Should we clear the element const inserted before returning?